File: C++/findShortestSubArray.cpp
/**
 * @File    : findShortestSubArray.cpp
 * @Brief   : 数组的度  
 * @Link    : https://leetcode-cn.com/problems/degree-of-an-array/
 * @Author  : Wei Li
 * @Date    : 2021-07-27
*/

#include <iostream>
#include <vector>
#include <unordered_map>

class Solution
{
public:
    int findShortestSubArray(std::vector<int> &nums)
    {
        std::unordered_map<int, std::vector<int>> mp;
        int n = nums.size();
        for (int i = 0; i < n; i++)
        {
            if (mp.count(nums[i]))
            {
                mp[nums[i]][0]++;
                mp[nums[i]][2] = i;
            }
            else
            {
                mp[nums[i]] = {1, i, i};
            }
        }
        int maxNum = 0, minLen = 0;
        for (auto &[_, vec] : mp)  // C++17
        {
            if (maxNum < vec[0])
            {
                maxNum = vec[0];
                minLen = vec[2] - vec[1] + 1;
            }
            else if (maxNum == vec[0])
            {
                if (minLen > vec[2] - vec[1] + 1)
                {
                    minLen = vec[2] - vec[1] + 1;
                }
            }
        }
        return minLen;
    }
};

// ----------------------------
int main(int argc, char **argv)
{
    std::vector<int> nums{1, 2, 2, 3, 1};

    Solution solution;
    auto subarray = solution.findShortestSubArray(nums);
    std::cout << "The solution of this problem is " << subarray << std::endl;

    return 0;
}

File: Python/findShortestSubArray.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: findShortestSubArray.py
@Function: 数组的度
@Link: https://leetcode-cn.com/problems/degree-of-an-array/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-07-27
"""


from typing import List


class Solution:
    def findShortestSubArray(self, nums: List[int]) -> int:
        mp = dict()

        for i, num in enumerate(nums):
            if num in mp:
                mp[num][0] += 1
                mp[num][2] = i
            else:
                mp[num] = [1, i, i]

        maxNum = minLen = 0
        for count, left, right in mp.values():
            if maxNum < count:
                maxNum = count
                minLen = right - left + 1
            elif maxNum == count:
                if minLen > (span := right - left + 1):
                    minLen = span

        return minLen


# -------------------------
if __name__ == "__main__":
    nums = [1,2,2,3,1,4,2]

    solution = Solution()
    subarray = solution.findShortestSubArray(nums)
    print(f"The solution of this problem is {subarray}")
